### ISSUE TITLE
Add custom scalar loss to 'categorical_cross_entropy'

### ISSUE THREAD

**Issue by lokesharma-dev** [tensorflow/tensorflow#41136]
I wish to add a custom loss to regular cross-entropy used when we build/compile a tf model using model.compile()
However, I get an error as provided in the below snapshot. 

Code:
def createEmbedding(features):
  seq = Input(shape=(300,))
  emb = Embedding(input_dim=len(vocab)+1,
                  output_dim = 50,
                  weights = [embedding_matrix],
                  trainable=False)(seq)
  emb_model = Model(seq, emb)
  clean_emb = emb_model(features)
  return clean_emb

def createModel(embedding_features):
  emb_tensor = Input(shape=(300,50,))
  hidden = LSTM(units=128)(emb_tensor)
  output = Dense(units=32, activation='relu')(hidden)
  model = Model(inputs=emb_tensor, outputs=output)
  logits = model(embedding_features)
  return emb_tensor, output, logits

def compute_kld(p_logit, q_logit):
  p = tf.nn.softmax(p_logit)
  q = tf.nn.softmax(q_logit)
  kl_score = tf.reduce_sum( p * (tf.math.log(p+1e-16) - tf.math.log(q+1e-16)), axis = 1)
  return kl_score

def calculateGradient(clean_features, noised_features):
  with tf.GradientTape(watch_accessed_variables=False) as tape:
    tape.watch(noised_features)
    _, _, p_logit = createModel(clean_features)
    _, _, p_logit_r = createModel(noised_features)
    kl_score = compute_kld(p_logit, p_logit_r)
    print(kl_score)
  grads = tape.gradient(kl_score, noised_features)
  return grads

features, labels = next(iter(train_dataset))

clean_features = createEmbedding(features)
noised_features = tf.add(clean_features, 0.01)

print('Clean Embedding: ', type(clean_features), clean_features.shape) # <class 'tensorflow.python.framework.ops.EagerTensor'> (1024, 300, 50)
print('Noised Embedding: ', type(noised_features), noised_features.shape) # <class 'tensorflow.python.framework.ops.EagerTensor'> (1024, 300, 50)

clean_ip_tensor, clean_op_tensor, p_logit = createModel(clean_features)
noise_ip_tensor, noise_op_tensor, p_logit_r = createModel(noised_features)
print('P_Logit: ',type(p_logit), p_logit.shape) # <class 'tensorflow.python.framework.ops.EagerTensor'> (1024, 32)
print('P_Logit_R: ',type(p_logit_r), p_logit_r.shape) # <class 'tensorflow.python.framework.ops.EagerTensor'> (1024, 32)

grads = calculateGradient(clean_features, noised_features)
norm_ball = tf.math.l2_normalize(grads, axis=None, epsilon=1e-12, name=None)
type(norm_ball), norm_ball.shape
rvadv = (grads/norm_ball) * -1
vadv_features = tf.add(clean_features, rvadv)

vat_ip_tensor, vat_op_tensor, q_logit = createModel(vadv_features)
vat_loss = compute_kld(p_logit, q_logit) # <tf.Tensor: shape=(1024,), dtype=float32; these will be scalar value
# I wish to add this vat_loss to regular categorical cross entropy loss

p = Dense(units=1, activation='softmax')(clean_op_tensor)
model = Model(inputs=clean_ip_tensor, outputs=p)
model.add_loss(vat_loss) # Error
model.summary()

# Future implementation
model.compile(optimizer='Adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(clean_features, labels)

![Screenshot 2020-07-06 at 22 16 46](https://user-images.githubusercontent.com/63437596/86637762-682e6900-bfd6-11ea-8e46-cfd329eed5e6.png)

Any suggestions on how can I work around this scenario. 
Thanks for your time and consideration

**Comment by Saduf2019** [comment]
@lokesharma-dev 
Please provide with indented code such that we can replicate the issue faced, we ran the code shared, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/e0fc9ec4cce70933d69d50db9300d081/untitled262.ipynb) for the same.
or if possible share a colab gist with the error faced.

**Comment by lokesharma-dev** [comment]
@Saduf2019 . Please find the colab link: 
[https://colab.research.google.com/drive/1kCh04Wb0w3O_0ikuYY_Q8OwTlwwM2Sh-?usp=sharing](url)

The data used for the above colab: 
[npy.zip](https://github.com/tensorflow/tensorflow/files/4886027/npy.zip)

What I attempt of doing is from this gist [https://gist.github.com/divamgupta/c778c17459c1f162e789560d5e0b2f0b](url). However, the example is explained in TensorFlow 1.x. Any suggestion on how to implement the same in version 2.x.

Thanks
